### ISSUE TITLE
Prevent invalid usage of Interceptors

### ISSUE THREAD

**Issue by Sean-Der** [pion/webrtc#1956]
Interceptors are being accidentally misused by users.
I have helped multiple users via GitHub and Slack debug
the same root cause with different behaviors. The issue
is that an Interceptor can be re-used between multiple
PeerConnections. Interceptors were designed to only be
single PeerConnection aware, so state is being corrupted.

The re-use looks like this.

```
m := &webrtc.MediaEngine{}
m.RegisterDefaultCodecs()
webrtc.RegisterDefaultInterceptors(m, i)

api := webrtc.NewAPI(webrtc.WithMediaEngine(m), webrtc.WithInterceptorRegistry(i))

a, _ := api.NewPeerConnection(config)
b, _ := api.NewPeerConnection(config)
c, _ := api.NewPeerConnection(config)

```

PeerConnection `a`, `b` and `c` all will use the same Interceptors. This will cause
Sender Reports, Receiver Reports, Transport Wide Congestion Control Feedback and NACKs
to be shared between them all.

----

These are some of the requirements.

* pion/webrtc breakage should be avoided, I don't want to cut v4 just for this
* We should support interceptors being scoped to one OR many PeerConnections.

We have three possible fixes. If anyone has different ideas I would
love to hear them!

### Add Copy() to Interceptors

Adds an extra method to each Interceptor. When we start each PeerConnection
we call copy.

### Add Sessions to Interceptors

Add sessionId to each function call in Interceptors. Allows an interceptor to be
multi-session aware.

### Return an error from NewPeerConnection on Interceptor re-usage

We can detect that an InterceptorRegistry has been built twice. This allows us to throw
an error and people can refactor their code

---- 

@jeremija opened an API for an implementation that added sessions https://github.com/pion/interceptor/pull/38

We also discussed this somewhat in https://github.com/pion/webrtc/issues/1749. I was focused on the wrong part of the issue so wanted to start a new discussion here.

**Comment by Sean-Der** [comment]
My vote is for `Copy()`

* Keeps Interceptors simple. They don't have to be session aware
* If users want to implement a 'Many Session' Interceptor `Copy()` can return a handle that gets added to the object getting copied.
* We don't need to break the API

**Comment by Sean-Der** [comment]
@at-wat @jeremija @jech @OrlandoCo @billylindeman @adamroach @davidzhao @mengelbart @zyxar 

Thoughts/opinions. You all have done work around Interceptors/Media API so would love your input on this.

**Comment by adamroach** [comment]
From the perspective of what makes Pion easier to develop on top of, I definitely support an approach that allows creation of multiple PeerConnections on top of the same interceptor registry without errors.

Among the three approaches described, I agree that `Copy()` is the most appealing; but I'd like to propose a fourth approach for consideration. Under the described scheme, you're creating prototypical instances of each interceptor that never actually get used. Among object creation strategies, usually you use prototypical instances only when creation is expensive but copying is cheap. For the interceptors we have right now, I don't _think_ that's the case

So I _think_ it would be cleaner if, instead of registering a set of interceptors, we instead register a set of interceptor factories that are invoked when a PeerConnection is created. To be clear, I haven't looked closely at this code, so there might be reasons this isn't ideal; but as a general software design strategy, that's the traditional approach.

But if that's not feasible for some reason, then the prototypical instance approach should work fine.

**Comment by davidzhao** [comment]
I think Copy works well in the sense that most users don't have to care that it exists. It could still cause confusion for folks developing interceptors, if they do want/expect some sort of shared state.

Another option is to make this all explicit by introducing a `InterceptorFactory` that is what the registry would use. This would make things much more explicit vs implicitly copying upon use.

**Comment by Sean-Der** [comment]
@adamroach @davidzhao 👍 for me with Factory approach! That sounds even better.

I am going to start a PR now and will send it over to you two hopefully in the next day or so. Thanks for the feedback/ideas

**Comment by at-wat** [comment]
> ## Add Sessions to Interceptors
>
>Add sessionId to each function call in Interceptors. Allows an interceptor to be multi-session aware.

My opinion about this: https://github.com/pion/webrtc/issues/1749#issuecomment-876104492

**Comment by jeremija** [comment]
I'm late to the party, but I agree with the factory approach. If we ever want to use a single event loop to process the RTP/RTCP packets of multiple peer connections, that should probably come at a later point, as an upgrade.

**Comment by Sean-Der** [comment]
Resolved by dce970438344727af9c9965f88d958c55d32e64d
